snuba/datasets/configuration/entity_builder.py

```diff
@@ -0,0 +1,130 @@
+from snuba.datasets.entity import Entity
+from snuba.datasets.pluggable_entity import PluggableEntity
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage, get_writable_storage
+from snuba.query.processors import QueryProcessor
+from snuba.query.validation.validators import (
+    EntityRequiredColumnValidator,
+    QueryValidator,
+)
+
+__MAPPINGS_HAVE_BEEN_INITIALIZED = False
+
+_QP_MAPPING: dict[str, Type[QueryProcessor]] = {}
+
+_FUNCTION_MAPPER_MAPPING: Mapping[str, Type[FunctionCallMapper]] = {
+    "simple_func": FunctionNameMapper
+}
+
+_SUB_MAPPER_MAPPING: Mapping[str, Type[SubscriptableReferenceMapper]] = {
+    "subscriptable": SubscriptableMapper
+}
+
+_VALIDATOR_MAPPING: Mapping[str, Type[QueryValidator]] = {
+    "entity_required": EntityRequiredColumnValidator
+}
+
+logger = logging.getLogger("snuba.entity_builder")
+
+
+def _initialize_mappings() -> None:
+    # circular imports 🤪
+    global __MAPPINGS_HAVE_BEEN_INITIALIZED
+    from snuba.datasets.entities.metrics import TagsTypeTransformer
+    from snuba.query.processors.granularity_processor import MappedGranularityProcessor
+    from snuba.query.processors.object_id_rate_limiter import (
+        OrganizationRateLimiterProcessor,
+        ProjectRateLimiterProcessor,
+        ProjectReferrerRateLimiter,
+        ReferrerRateLimiterProcessor,
+    )
+    from snuba.query.processors.quota_processor import ResourceQuotaProcessor
+    from snuba.query.processors.timeseries_processor import TimeSeriesProcessor
+
+    _QP_MAPPING.update(
+        {
+            "transform_tag_types": TagsTypeTransformer,
+            "handle_mapped_granularities": MappedGranularityProcessor,
+            "translate_time_series": TimeSeriesProcessor,
+            "referrer_rate_limit": ReferrerRateLimiterProcessor,
+            "org_rate_limiter": OrganizationRateLimiterProcessor,
+            "project_referrer_rate_limiter": ProjectReferrerRateLimiter,
+            "project_rate_limiter": ProjectRateLimiterProcessor,
+            "resource_quota_limiter": ResourceQuotaProcessor,
+        }
+    )
+
+    __MAPPINGS_HAVE_BEEN_INITIALIZED = True
+
+
+def _build_entity_validators(
+    config_validators: list[dict[str, Any]]
+) -> Sequence[QueryValidator]:
+    return [
+        _VALIDATOR_MAPPING[qv_config["validator"]](**qv_config["args"])  # type: ignore
+        for qv_config in config_validators
+    ]
+
+
+def _build_entity_query_processors(
+    config_query_processors: list[dict[str, Any]],
+) -> Sequence[QueryProcessor]:
+    return [_QP_MAPPING[config_qp["processor"]](**(config_qp["args"] if config_qp.get("args") else {})) for config_qp in config_query_processors]  # type: ignore
+
+
+def _build_entity_translation_mappers(
+    config_translation_mappers: dict[str, Any],
+) -> TranslationMappers:
+    function_mappers: list[FunctionCallMapper] = [
+        _FUNCTION_MAPPER_MAPPING[fm_config["mapper"]](**fm_config["args"])  # type: ignore
```

Comment:
I understand we can't actually type these, but I would rather have loose typing than the ignore comment. We could achieve that (I think) by adding a `def __init__(self, *args: Any, **kwargs: Any)` to the `QueryProcessor` class.

Reply:
This is actually already an issue, see my comment above about the `EntityRequiredColumnValidator`. Do we need to change processors to strictly validate their args? This feels like a larger discussion.

Reply:
> This is actually already an issue, see my comment above about the `EntityRequiredColumnValidator`. Do we need to change processors to strictly validate their args? This feels like a larger discussion.

Maybe this could be incorporated into the registry/factory pattern where in addition to a friendly config name you also have to describe the method signature types? Maybe this could be done with introspection? Or it might be as simple as a practice of 
```
assert isinstance(arg1, type1) 
assert isinstance(arg2, type2) 
```

though that is pretty manual and error-prone

Reply:
I tried doing what you suggested with a method body of `raise NotImplementedError` but I it breaks the cases where there is no initializer defined on the subclass (see `TagsTypeTransformer`, `TagsExpanderTransformer`). Would you want to just have an __init__ with a pass as the method body?

Reply:
After our discussion, let's leave this as is in this PR. There's a larger task here to figure out how to validate arguments in these situations that I will make a separate ticket for.